addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,423 @@
+    keyOptions.lid[newLid] = identifier;
+    // ensure a peekAll sees our new identifier too
+    // TODO move this outta here?
+    keyOptions._allIdentifiers.push(identifier);
+
+    return identifier;
+  }
+
+  /*
+   Provides the opportunity to update secondary lookup tables for existing identifiers
+   Called after an identifier created with `createIdentifierForNewRecord` has been
+   committed.
+
+   Assigned `id` to an `Identifier` if `id` has not previously existed; however,
+   attempting to change the `id` or calling update without providing an `id` when
+   one is missing will throw an error.
+
+    - sets `id` (if `id` was previously `null`)
+    - `lid` and `type` MUST NOT be altered post creation
+  */
+  updateRecordIdentifier(identifier: StableRecordIdentifier, data: ResourceIdentifierObject): void {
+    if (DEBUG) {
+        assert(
+          `The supplied identifier '${identifier}' does not belong to this store instance.`,
+           this._cache.lids[identifier.lid] === identifier
+        );
+    
+        let id = identifier.id;
+        let newId = coerceId(data.id);
+    
+        if (id !== null && id !== newId && newId !== null) {
+          let keyOptions = getTypeIndex(this._cache.types, identifier.type);
+          let existingIdentifier = keyOptions.id[newId];
+
+          if (existingIdentifier !== undefined) {
+            throw new Error(
+              `Attempted to update the 'id' for the RecordIdentifier '${identifier}' to '${newId}', but that id is already in use by '${existingIdentifier}'`
+            );
+          }
+        }
+      }
+    
+      let id = identifier.id;
+      performRecordIdentifierUpdate(identifier, data, this._update);
+      let newId = identifier.id;
+    
+      // add to our own secondary lookup table
+      if (id !== newId && newId !== null) {
+        let keyOptions = getTypeIndex(this._cache.types, identifier.type);
+        keyOptions.id[newId] = identifier;
+      }
+  }
+
+  /*
+   Provides the opportunity to eliminate an identifier from secondary lookup tables
+   as well as eliminates it from ember-data's own lookup tables and book keeping.
+
+   Useful when a record has been deleted and the deletion has been persisted and
+   we do not care about the record anymore. Especially useful when an `id` of a
+   deleted record might be reused later for a new record.
+  */
+  forgetRecordIdentifier(identifier: StableRecordIdentifier): void {
+    let keyOptions = getTypeIndex(this._cache.types, identifier.type);
+    if (identifier.id !== null) {
+      delete keyOptions.id[identifier.id];
+    }
+    delete this._cache.lids[identifier.lid];
+    delete keyOptions.lid[identifier.lid];
+  
+    let index = keyOptions._allIdentifiers.indexOf(identifier);
+    keyOptions._allIdentifiers.splice(index, 1);
+
+    this._forget(identifier, 'record');
+  }
+}
+
+function isNonEmptyString(str?: string | null): str is string {
+  return typeof str === 'string' && str.length > 0;
+}
```

Comment:
You can make this a more useful check through using a generic.
```ts
function isNonEmptyString<T extends string>(str: T | undefined | null | ''): str is T {
  return typeof str === 'string' && str.length > 0;
}
```

This will allow it to work correctly with unioned string literal types

```ts
let x = 'bar' as '' | 'foo' | 'bar';

if (isNonEmptyString(x)) {
    x // 'foo' | 'bar'  (yours would have included '')
} else {
    x // '' (yours would have been never)
}

```